Sources/ProcedureKit/Procedure.swift

```diff
@@ -100,44 +122,69 @@ open class Procedure: Operation, ProcedureProtocol {
         }
     }
 
+    @available(OSX 10.10, iOS 8.0, tvOS 8.0, watchOS 2.0, *)
+    open override var qualityOfService: QualityOfService {
+        get { return super.qualityOfService }
+        set {
+            super.qualityOfService = newValue
+            eventQueue.qualityOfService = newValue.qos
+        }
+    }
+
     internal let identifier = UUID()
 
+    internal class ProcedureQueueContext { }
+    internal let queueAddContext = ProcedureQueueContext()
+
+    deinit {
+        // ensure that the Protected Properies are deinitialized within the lock
+        _stateLock.withCriticalScope {
+            self.protectedProperties = nil
+        }
+    }
+
     // MARK: State
 
-    private var _state = ProcedureKit.State.initialized
-    private let _stateLock = NSRecursiveLock()
+    // the state variable to be used *within* the stateLock
+    private var _state = ProcedureKit.State.initialized {
+        willSet(newState) {
+            _log.verbose(message: "\(_state) -> \(newState)")
+            assert(_state.canTransition(to: newState, whenCancelled: _isCancelled), "Attempting to perform illegal cyclic state transition, \(_state) -> \(newState) for operation: \(identity). Ensure that Procedure instances are added to a ProcedureQueue not an OperationQueue.")
+        }
+    }
+
+    fileprivate let _stateLock = PThreadMutex()
```

Comment:
Exposing the Procedure's internal _stateLock is very, very problematic (very easy to introduce deadlock if people don't know what they're doing), but this touches the part I haven't written up yet for the documentation:

You can use the Procedure's EventQueue. (See `Procedure.eventQueue`). You can simply dispatch blocks onto the EventQueue to do things that manipulate state.